File: apps/visualstudio/ResXGuard.Core/Services/ResxText.cs
using System.Globalization;
using System.Text;
using System.Text.RegularExpressions;

namespace ResXGuard.Core;

/// <summary>
/// Surgical .resx text edits so encoding, newlines, and the XML declaration stay intact.
/// </summary>
public static class ResxText
{
    private static readonly Regex DataOpen = new(@"<data\b", RegexOptions.IgnoreCase | RegexOptions.Compiled);
    private static readonly Regex NameAttr = new(@"\bname\s*=\s*""([^""]*)""", RegexOptions.IgnoreCase | RegexOptions.Compiled);
    private static readonly Regex RootClose = new(@"</root\s*>", RegexOptions.IgnoreCase | RegexOptions.Compiled);

    public static string DetectNewline(string text)
    {
        var crlf = 0;
        var lf = 0;
        for (var i = 0; i < text.Length; i++)
        {
            if (text[i] != '\n') continue;
            if (i > 0 && text[i - 1] == '\r') crlf++;
            else lf++;
        }
        return crlf > lf ? "\r\n" : "\n";
    }

    public static string EscapeXmlText(string value) =>
        value.Replace("&", "&amp;").Replace("<", "&lt;").Replace(">", "&gt;");

    public static string EscapeXmlAttr(string value) =>
        EscapeXmlText(value).Replace("\"", "&quot;");

    public static int CompareResxKeys(string a, string b)
    {
        var cmp = CompareInfo.InvariantInfo.Compare(
            a,
            b,
            CompareOptions.IgnoreCase | CompareOptions.IgnoreNonSpace);
        return cmp != 0 ? cmp : string.CompareOrdinal(a, b);
    }

    public static string RenameResxKeyInXml(string xml, string oldKey, string newKey)
    {
        var block = FindBlock(xml, oldKey);
        if (block == null) return xml;
        return SortResxDataEntries(ReplaceRange(xml, block.NameValueStart, block.NameValueEnd, EscapeXmlAttr(newKey)));
    }

    public static string SetResxValueInXml(string xml, string key, string value, string? comment = null)
    {
        var block = FindBlock(xml, key);
        if (block == null)
            return InsertDataBlock(xml, key, value, comment);

        var next = xml;
        var escaped = EscapeXmlText(value);
        if (block.ValueInnerStart >= 0)
            next = ReplaceRange(next, block.ValueInnerStart, block.ValueInnerEnd, escaped);
        else
        {
            var newline = DetectNewline(xml);
            var indent = IndentOf(xml, block.Start);
            next = ReplaceRange(next, block.OpenTagEnd, block.OpenTagEnd, $"{newline}{indent}  <value>{escaped}</value>");
        }

        if (comment != null)
            next = SetResxCommentInXml(next, key, comment);
        return next;
    }

    public static string SetResxCommentInXml(string xml, string key, string comment)
    {
        var block = FindBlock(xml, key);
        if (block == null) return xml;
        var escaped = EscapeXmlText(comment);
        if (block.HasComment && block.CommentInnerStart >= 0)
            return ReplaceRange(xml, block.CommentInnerStart, block.CommentInnerEnd, escaped);

        var newline = DetectNewline(xml);
        var indent = IndentOf(xml, block.Start);
        var insertAt = block.OpenTagEnd;
        if (block.ValueInnerEnd >= 0)
        {
            var valueClose = IndexOfIgnoreCase(xml, "</value>", block.ValueInnerEnd);
            if (valueClose >= 0) insertAt = valueClose + "</value>".Length;
        }
        return ReplaceRange(xml, insertAt, insertAt, $"{newline}{indent}  <comment>{escaped}</comment>");
    }

    public static string DeleteResxEntryInXml(string xml, string key)
    {
        var block = FindBlock(xml, key);
        if (block == null) return xml;
        var start = block.Start;
        var end = block.End;
        while (start > 0 && (xml[start - 1] == ' ' || xml[start - 1] == '\t'))
            start--;
        if (end < xml.Length - 1 && xml[end] == '\r' && xml[end + 1] == '\n')
            end += 2;
        else if (end < xml.Length && xml[end] == '\n')
            end += 1;
        return xml.Substring(0, start) + xml.Substring(end);
    }

    public static string AddResxEntryInXml(string xml, string key, string value, string comment = "")
    {
        if (FindBlock(xml, key) != null)
            return SetResxValueInXml(xml, key, value, string.IsNullOrEmpty(comment) ? null : comment);
        return InsertDataBlock(xml, key, value, comment);
    }

    private static string InsertDataBlock(string xml, string key, string value, string? comment)
    {
        var newline = DetectNewline(xml);
        var blocks = FindRootDataBlocks(xml);
        var indent = blocks.Count > 0 ? IndentOf(xml, blocks[0].Start) : "  ";
        var commentXml = string.IsNullOrEmpty(comment)
            ? ""
            : $"{newline}{indent}  <comment>{EscapeXmlText(comment)}</comment>";
        var block =
            $"{indent}<data name=\"{EscapeXmlAttr(key)}\" xml:space=\"preserve\">{newline}{indent}  <value>{EscapeXmlText(value)}</value>{commentXml}{newline}{indent}</data>";
        var closeAt = RootCloseIndex(xml);
        var before = xml.Substring(0, closeAt);
        var needsNl = before.EndsWith("\n", StringComparison.Ordinal) ? "" : newline;
        return SortResxDataEntries(before + needsNl + block + newline + xml.Substring(closeAt));
    }

    public static string SortResxDataEntries(string xml)
    {
        var blocks = FindRootDataBlocks(xml);
        if (blocks.Count < 2) return xml;
        var spans = blocks.Select(block =>
        {
            var (start, end) = SpanOfBlock(xml, block);
            return (block.Key, start, end, text: xml.Substring(start, end - start));
        }).ToList();
        for (var i = 1; i < spans.Count; i++)
        {
            if (!string.IsNullOrWhiteSpace(xml.Substring(spans[i - 1].end, spans[i].start - spans[i - 1].end)))
                return xml;
        }
        var sorted = spans.OrderBy(s => s.Key, Comparer<string>.Create(CompareResxKeys)).ToList();
        if (sorted.Select((s, i) => s.Key == spans[i].Key).All(eq => eq))
            return xml;
        return xml.Substring(0, spans[0].start)
               + string.Concat(sorted.Select(s => s.text))
               + xml.Substring(spans[^1].end);
    }

    private static (int start, int end) SpanOfBlock(string xml, DataBlock block)
    {
        var start = block.Start;
        var end = block.End;
        while (start > 0 && (xml[start - 1] == ' ' || xml[start - 1] == '\t'))
            start--;
        if (end < xml.Length - 1 && xml[end] == '\r' && xml[end + 1] == '\n')
            end += 2;
        else if (end < xml.Length && xml[end] == '\n')
            end += 1;
        return (start, end);
    }

    private sealed class DataBlock
    {
        public int Start { get; init; }
        public int End { get; init; }
        public int OpenTagEnd { get; init; }
        public string Key { get; init; } = "";
        public int NameValueStart { get; init; }
        public int NameValueEnd { get; init; }
        public int ValueInnerStart { get; init; }
        public int ValueInnerEnd { get; init; }
        public int CommentInnerStart { get; init; }
        public int CommentInnerEnd { get; init; }
        public bool HasComment { get; init; }
    }

    private static DataBlock? FindBlock(string xml, string key) =>
        FindRootDataBlocks(xml).FirstOrDefault(b => b.Key == key);

    private static List<DataBlock> FindRootDataBlocks(string xml)
    {
        var blocks = new List<DataBlock>();
        var offset = 0;
        while (true)
        {
            var match = DataOpen.Match(xml, offset);
            if (!match.Success) break;
            var start = match.Index;
            var openTagEnd = xml.IndexOf('>', start);
            if (openTagEnd < 0) break;
            var openTag = xml.Substring(start, openTagEnd + 1 - start);
            if (Regex.IsMatch(openTag, @"/\s*>$"))
            {
                offset = openTagEnd + 1;
                continue;
            }
            var nameMatch = NameAttr.Match(openTag);
            if (!nameMatch.Success)
            {
                offset = openTagEnd + 1;
                continue;
            }
            var valueStartInTag = nameMatch.Index + nameMatch.Value.IndexOf('"') + 1;
            var closeAt = IndexOfIgnoreCase(xml, "</data>", openTagEnd + 1);
            if (closeAt < 0) break;
            var end = closeAt + "</data>".Length;
            var value = FindInnerElement(xml, openTagEnd + 1, closeAt, "value");
            var comment = FindInnerElement(xml, openTagEnd + 1, closeAt, "comment");
            blocks.Add(new DataBlock
            {
                Start = start,
                End = end,
                OpenTagEnd = openTagEnd + 1,
                Key = nameMatch.Groups[1].Value,
                NameValueStart = start + valueStartInTag,
                NameValueEnd = start + valueStartInTag + nameMatch.Groups[1].Length,
                ValueInnerStart = value?.Start ?? -1,
                ValueInnerEnd = value?.End ?? -1,
                CommentInnerStart = comment?.Start ?? -1,
                CommentInnerEnd = comment?.End ?? -1,
                HasComment = comment != null
            });
            offset = end;
        }
        return blocks;
    }

    private static (int Start, int End)? FindInnerElement(string xml, int innerStart, int innerEnd, string tag)
    {
        var open = new Regex($"<{tag}\\b[^>]*>", RegexOptions.IgnoreCase);
        var slice = xml.Substring(innerStart, innerEnd - innerStart);
        var openMatch = open.Match(slice);
        if (!openMatch.Success) return null;
        var contentStart = innerStart + openMatch.Index + openMatch.Length;
        var closeAt = IndexOfIgnoreCase(xml, $"</{tag}>", contentStart);
        if (closeAt < 0 || closeAt > innerEnd) return null;
        return (contentStart, closeAt);
    }

    private static string IndentOf(string xml, int index)
    {
        var i = index;
        while (i > 0 && xml[i - 1] != '\n') i--;
        var prefix = xml.Substring(i, index - i);
        return string.IsNullOrWhiteSpace(prefix) ? prefix : "  ";
    }

    private static int RootCloseIndex(string xml)
    {
        var match = RootClose.Match(xml);
        return match.Success ? match.Index : xml.Length;
    }

    private static string ReplaceRange(string xml, int start, int end, string insert) =>
        xml.Substring(0, start) + insert + xml.Substring(end);

    private static int IndexOfIgnoreCase(string haystack, string needle, int startIndex)
    {
        return haystack.IndexOf(needle, startIndex, StringComparison.OrdinalIgnoreCase);
    }
}

File: packages/core-ts/src/services/resx-text.ts
/**
 * Surgical .resx text edits. Parse with the DOM for reads; mutate the original
 * XML string for writes so encoding, newlines, declaration, and schema stay put.
 */

export type ResxNewline = '\n' | '\r\n';

export function detectNewline(text: string): ResxNewline {
  let crlf = 0;
  let lf = 0;
  for (let i = 0; i < text.length; i++) {
    if (text[i] === '\n') {
      if (i > 0 && text[i - 1] === '\r') {
        crlf++;
      } else {
        lf++;
      }
    }
  }
  return crlf > lf ? '\r\n' : '\n';
}

export function escapeXmlText(value: string): string {
  return value.replace(/&/g, '&amp;').replace(/</g, '&lt;').replace(/>/g, '&gt;');
}

export function escapeXmlAttr(value: string): string {
  return escapeXmlText(value).replace(/"/g, '&quot;');
}

interface DataBlock {
  start: number;
  end: number;
  openTagStart: number;
  openTagEnd: number;
  key: string;
  nameValueStart: number;
  nameValueEnd: number;
  valueInnerStart: number;
  valueInnerEnd: number;
  commentInnerStart: number;
  commentInnerEnd: number;
  hasComment: boolean;
}

const DATA_OPEN = /<data\b/gi;

function findNameAttr(openTag: string, absStart: number): { key: string; valueStart: number; valueEnd: number } | null {
  const match = /\bname\s*=\s*"([^"]*)"/i.exec(openTag);
  if (!match || match.index === undefined) {
    return null;
  }
  const valueStart = absStart + match.index + match[0].indexOf('"') + 1;
  return {
    key: match[1],
    valueStart,
    valueEnd: valueStart + match[1].length,
  };
}

function findInnerElement(
  xml: string,
  innerStart: number,
  innerEnd: number,
  tag: 'value' | 'comment'
): { innerStart: number; innerEnd: number } | null {
  const open = new RegExp(`<${tag}\\b[^>]*>`, 'i');
  const slice = xml.slice(innerStart, innerEnd);
  const openMatch = open.exec(slice);
  if (!openMatch) {
    return null;
  }
  const contentStart = innerStart + openMatch.index + openMatch[0].length;
  const closeTag = `</${tag}>`;
  const closeAt = xml.toLowerCase().indexOf(closeTag, contentStart);
  if (closeAt < 0 || closeAt > innerEnd) {
    return null;
  }
  return { innerStart: contentStart, innerEnd: closeAt };
}

export function findRootDataBlocks(xml: string): DataBlock[] {
  const blocks: DataBlock[] = [];
  DATA_OPEN.lastIndex = 0;
  let match: RegExpExecArray | null;
  while ((match = DATA_OPEN.exec(xml))) {
    const start = match.index;
    const openTagEnd = xml.indexOf('>', start);
    if (openTagEnd < 0) {
      break;
    }
    const openTag = xml.slice(start, openTagEnd + 1);
    if (/\/\s*>$/.test(openTag)) {
      DATA_OPEN.lastIndex = openTagEnd + 1;
      continue;
    }
    const name = findNameAttr(openTag, start);
    if (!name) {
      DATA_OPEN.lastIndex = openTagEnd + 1;
      continue;
    }
    const closeAt = xml.toLowerCase().indexOf('</data>', openTagEnd + 1);
    if (closeAt < 0) {
      break;
    }
    const end = closeAt + '</data>'.length;
    const value = findInnerElement(xml, openTagEnd + 1, closeAt, 'value');
    const comment = findInnerElement(xml, openTagEnd + 1, closeAt, 'comment');
    blocks.push({
      start,
      end,
      openTagStart: start,
      openTagEnd: openTagEnd + 1,
      key: name.key,
      nameValueStart: name.valueStart,
      nameValueEnd: name.valueEnd,
      valueInnerStart: value?.innerStart ?? -1,
      valueInnerEnd: value?.innerEnd ?? -1,
      commentInnerStart: comment?.innerStart ?? -1,
      commentInnerEnd: comment?.innerEnd ?? -1,
      hasComment: Boolean(comment),
    });
    DATA_OPEN.lastIndex = end;
  }
  return blocks;
}

function replaceRange(xml: string, start: number, end: number, insert: string): string {
  return xml.slice(0, start) + insert + xml.slice(end);
}

function findBlock(xml: string, key: string): DataBlock | undefined {
  return findRootDataBlocks(xml).find((block) => block.key === key);
}

function indentOf(xml: string, index: number): string {
  let i = index;
  while (i > 0 && xml[i - 1] !== '\n') {
    i--;
  }
  const prefix = xml.slice(i, index);
  return /^\s*$/.test(prefix) ? prefix : '  ';
}

function rootCloseIndex(xml: string): number {
  const match = /<\/root\s*>/i.exec(xml);
  return match ? match.index : xml.length;
}

function formatDataElement(
  newline: ResxNewline,
  indent: string,
  key: string,
  value: string,
  comment?: string
): string {
  const inner = `${indent}  `;
  const commentXml =
    comment !== undefined && comment !== ''
      ? `${newline}${inner}<comment>${escapeXmlText(comment)}</comment>`
      : '';
  return `${indent}<data name="${escapeXmlAttr(key)}" xml:space="preserve">${newline}${inner}<value>${escapeXmlText(value)}</value>${commentXml}${newline}${indent}</data>`;
}

export function compareResxKeys(a: string, b: string): number {
  return a.localeCompare(b, undefined, { sensitivity: 'base', numeric: true }) || a.localeCompare(b);
}

export function renameResxKeyInXml(xml: string, oldKey: string, newKey: string): string {
  const block = findBlock(xml, oldKey);
  if (!block) {
    return xml;
  }
  const next = replaceRange(xml, block.nameValueStart, block.nameValueEnd, escapeXmlAttr(newKey));
  return sortResxDataEntries(next);
}

export function setResxValueInXml(xml: string, key: string, value: string, comment?: string): string {
  const block = findBlock(xml, key);
  const escaped = escapeXmlText(value);
  if (!block) {
    return insertDataBlock(xml, key, value, comment);
  }
  let next = xml;
  if (block.valueInnerStart >= 0) {
    next = replaceRange(next, block.valueInnerStart, block.valueInnerEnd, escaped);
  } else {
    const insertAt = block.openTagEnd;
    const newline = detectNewline(xml);
    const indent = indentOf(xml, block.start);
    next = replaceRange(next, insertAt, insertAt, `${newline}${indent}  <value>${escaped}</value>`);
  }
  if (comment !== undefined) {
    next = setResxCommentInXml(next, key, comment);
  }
  return next;
}

export function setResxCommentInXml(xml: string, key: string, comment: string): string {
  const block = findBlock(xml, key);
  if (!block) {
    return xml;
  }
  const escaped = escapeXmlText(comment);
  if (block.hasComment && block.commentInnerStart >= 0) {
    return replaceRange(xml, block.commentInnerStart, block.commentInnerEnd, escaped);
  }
  const newline = detectNewline(xml);
  const indent = indentOf(xml, block.start);
  let insertAt = block.openTagEnd;
  if (block.valueInnerEnd >= 0) {
    const valueClose = xml.toLowerCase().indexOf('</value>', block.valueInnerEnd);
    if (valueClose >= 0) {
      insertAt = valueClose + '</value>'.length;
    }
  }
  return replaceRange(xml, insertAt, insertAt, `${newline}${indent}  <comment>${escaped}</comment>`);
}

export function deleteResxEntryInXml(xml: string, key: string): string {
  const block = findBlock(xml, key);
  if (!block) {
    return xml;
  }
  let start = block.start;
  let end = block.end;
  while (start > 0 && (xml[start - 1] === ' ' || xml[start - 1] === '\t')) {
    start--;
  }
  if (end < xml.length && xml[end] === '\r' && xml[end + 1] === '\n') {
    end += 2;
  } else if (end < xml.length && xml[end] === '\n') {
    end += 1;
  }
  return xml.slice(0, start) + xml.slice(end);
}

function insertDataBlock(xml: string, key: string, value: string, comment?: string): string {
  const newline = detectNewline(xml);
  const blocks = findRootDataBlocks(xml);
  const indent = blocks.length > 0 ? indentOf(xml, blocks[0].start) : '  ';
  const block = formatDataElement(newline, indent, key, value, comment);
  const closeAt = rootCloseIndex(xml);
  const before = xml.slice(0, closeAt);
  const needsNl = before.endsWith('\n') ? '' : newline;
  return sortResxDataEntries(`${before}${needsNl}${block}${newline}${xml.slice(closeAt)}`);
}

/** Reorder contiguous `<data>` entries A–Z. Schema and resheaders stay put. */
export function sortResxDataEntries(xml: string): string {
  const blocks = findRootDataBlocks(xml);
  if (blocks.length < 2) {
    return xml;
  }
  const spans = blocks.map((block) => {
    const { start, end } = spanOfBlock(xml, block);
    return { key: block.key, start, end, text: xml.slice(start, end) };
  });
  for (let i = 1; i < spans.length; i++) {
    if (xml.slice(spans[i - 1].end, spans[i].start).trim() !== '') {
      return xml;
    }
  }
  const sorted = [...spans].sort((a, b) => compareResxKeys(a.key, b.key));
  if (sorted.every((span, i) => span.key === spans[i].key)) {
    return xml;
  }
  return xml.slice(0, spans[0].start) + sorted.map((span) => span.text).join('') + xml.slice(spans[spans.length - 1].end);
}

function spanOfBlock(xml: string, block: DataBlock): { start: number; end: number } {
  let start = block.start;
  let end = block.end;
  while (start > 0 && (xml[start - 1] === ' ' || xml[start - 1] === '\t')) {
    start--;
  }
  if (end < xml.length && xml[end] === '\r' && xml[end + 1] === '\n') {
    end += 2;
  } else if (end < xml.length && xml[end] === '\n') {
    end += 1;
  }
  return { start, end };
}

export function addResxEntryInXml(xml: string, key: string, value: string, comment = ''): string {
  if (findBlock(xml, key)) {
    return setResxValueInXml(xml, key, value, comment || undefined);
  }
  return insertDataBlock(xml, key, value, comment);
}
